micro-rdk-nmea/src/parse_helpers/parsers.rs

```diff
@@ -0,0 +1,411 @@
+        }
+    }
+
+    pub fn advance(&self, bits: usize) -> Result<(), NmeaParseError> {
+        let advanced_position = self.bit_position.fetch_add(bits, Ordering::SeqCst);
+        if advanced_position > (self.data.len() * 8) {
+            self.bit_position.fetch_sub(bits, Ordering::SeqCst);
+            Err(NmeaParseError::EndOfBufferExceeded)
+        } else {
+            Ok(())
+        }
+    }
+}
+
+/// Trait for reading a data type (`FieldType`) from a DataCursor.
+pub trait FieldReader {
+    type FieldType;
+    fn read_from_cursor(&self, cursor: &DataCursor) -> Result<Self::FieldType, NmeaParseError>;
+}
+
+/// A field reader for parsing a basic number type. A reader with bit_size n will read its value
+/// as the first n bits of `size_of::<T>()` bytes with the remaining bits as zeroes (the resulting bytes
+/// will be parsed as Little-Endian). See invocation of the generate_number_field_readers macro below
+/// for the currently supported number types.
+pub struct NumberField<T> {
+    bit_size: usize,
+    _marker: PhantomData<T>,
+}
+
+impl<T> NumberField<T> {
+    pub fn new(bit_size: usize) -> Result<Self, NumberFieldError> {
+        let max_bit_size = std::mem::size_of::<T>() * 8;
+        if bit_size > max_bit_size {
+            Err(NumberFieldError::ImproperBitSize(bit_size, max_bit_size))
+        } else {
+            Ok(Self {
+                bit_size,
+                _marker: Default::default(),
+            })
+        }
+    }
+}
+
+impl<T> Default for NumberField<T>
+where
+    T: Sized,
+{
+    fn default() -> Self {
+        Self {
+            bit_size: size_of::<T>() * 8,
+            _marker: Default::default(),
+        }
+    }
+}
+
+macro_rules! generate_number_field_readers {
+    ($($t:ty),*) => {
+        $(
+            impl TryFrom<DataRead<'_>> for $t {
+                type Error = NumberFieldError;
+                fn try_from(value: DataRead) -> Result<Self, Self::Error> {
+                    let max_size = std::mem::size_of::<Self>();
+                    if (value.bit_size / 8) > max_size {
+                        return Err(NumberFieldError::ImproperBitSize(value.bit_size, max_size * 8));
+                    }
+
+                    let bits = &value.data[..].view_bits::<Msb0>();
+
+                    let start_idx = value.bit_position.load(Ordering::SeqCst) % 8;
+                    let end_idx = start_idx + value.bit_size;
+                    let mut bit_vec = bits[start_idx..end_idx].to_bitvec();
+                    if bit_vec.len() != (max_size * 8) {
+                        // the last byte is incomplete, we reverse the remaining bits of the
+                        // incomplete byte so it can be padded with zeros, then we reverse the bits
+                        // of the now complete last byte for the proper bit order
+                        let last_bit_start = bit_vec.len() - (value.bit_size % 8);
+                        let _ = &bit_vec[last_bit_start..].reverse();
```

Comment:
what is the purpose of this operation? for example if I  want to read 12bit on a 16bit number then last_bit_start would be 8 so the reverse operation will affect bits [8..12]. I think this ends up doing something like shift the left most bits right overwriting on the way. 
Also different value of bit_size will change the amount of bits reversed can that affect reading numbers?

Reply:
That is what I'm trying to do, yes. The reversing is just to be able to pad the bits with 0 from the front. So, if the overflow bits are 1110, I want to make it so the last byte is 00001110. To do that I reverse just the last 4 bits in place (0111), add the padding (01110000), then reverse the last byte to restore the bit order (00001110)

Reply:
are suggesting I could just right shift the last byte by however many bits are remaining?

Reply:
Not really. In my interpretation of the code (i think it's true for both version) if we have a data like this" `let data : [u8;2] = [0x0,0x10]` Then using a `NumberField<u16>` to read it will yield different result depending on whether we read 12,13,14,15 or 16 bits. This is because the amount of shifted bytes depend on the number of bits read, so the second byte of the array will be shifted differently like `0b010000 -> 0b000010` if bit read is 13 (shift 3) or like ` 0b010000 -> 0b001000` if bit read are 14 (shift 2). If this is how NMEA fields are parsed then we should just have an explanation so the code can be interpreted properly

Reply:
This is how the gonmea library appears to do it, should I try double-checking the canboat C code?